File: ITEM9 centroid_blob/multiple_blob.py
#coding=utf-8
import cv2
import numpy as np


img = cv2.imread('./image/multiple.png')

# convert the image to grayscale
gray_image = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
# convert the grayscale image to binary image
ret,thresh = cv2.threshold(gray_image,127,255,0)

# find contour in the binary image
contours, hierarchy = cv2.findContours(thresh,cv2.RETR_TREE,cv2.CHAIN_APPROX_SIMPLE)
for c in contours:
	# calculate moments for each contour
	M = cv2.moments(c)
	cX = int(M["m10"] / M["m00"])
	cY = int(M["m01"] / M["m00"])
	
	
    # calculate x,y coordinate of center
	cv2.circle(img, (cX, cY), 5, (255, 255, 255), -1)
	cv2.putText(img, "centroid", (cX - 25, cY - 25),cv2.FONT_HERSHEY_SIMPLEX, 0.5, (255, 255, 255), 2)


# 3.4.1 im2, contours, hierarchy = cv.findContours(thresh, cv.RETR_TREE, cv.CHAIN_APPROX_SIMPLE)
# 3.2.0 im2, contours, hierarchy = cv2.findContours(thresh,cv2.RETR_TREE,cv2.CHAIN_APPROX_SIMPLE) 	

# display the image
cv2.imshow("Image", img)
cv2.waitKey(0)

File: ITEM 1-9/ITEM9 centroid_blob/multiple_blob.cpp
#include "pch.h"
#include <opencv2/opencv.hpp>
#include <iostream>
#include <stdio.h>
#include <stdlib.h>

using namespace cv;
using namespace std;

RNG rng(12345);

void find_moments(Mat src);

int main()
{
	String img_path = "./image/multiple.png";
	/// Load source image, convert it to gray
	Mat src, gray;
	src = imread(img_path);

	cvtColor(src, gray, COLOR_BGR2GRAY);

	//显示原图
	namedWindow("Source", WINDOW_AUTOSIZE);
	imshow("Source", src);

	// call function to find_moments 寻质心函数
	find_moments(gray);

	waitKey(0);
	return(0);
}

void find_moments(Mat gray)
{
	Mat canny_output;
	//各个轮廓的点集合
	vector<vector<Point> > contours;
	//轮廓输出结果向量
	vector<Vec4i> hierarchy;

	/// Detect edges using canny 边缘算子提取轮廓
	Canny(gray, canny_output, 50, 150, 3);
	// Find contours 寻找轮廓 RETR_TREE表示提取所有轮廓
	findContours(canny_output, contours, hierarchy, RETR_TREE, CHAIN_APPROX_SIMPLE, Point(0, 0));

	/// Get the moments 图像矩
	vector<Moments> mu(contours.size());
	//求取每个轮廓的矩
	for (int i = 0; i < contours.size(); i++)
	{
		mu[i] = moments(contours[i], false);
	}

	///  Get the centroid of figures. 轮廓质点
	vector<Point2f> mc(contours.size());
	for (int i = 0; i < contours.size(); i++)
	{
		mc[i] = Point2f(mu[i].m10 / mu[i].m00, mu[i].m01 / mu[i].m00);
	}

	/// Draw contours
	//画轮廓
	Mat drawing(canny_output.size(), CV_8UC3, Scalar(255, 255, 255));

	for (int i = 0; i < contours.size(); i++)
	{
		Scalar color = Scalar(167, 151, 0);
		//画轮廓
		drawContours(drawing, contours, i, color, 2, 8, hierarchy, 0, Point());
		//画质心
		circle(drawing, mc[i], 4, color, -1, 7, 0);
	}

	/// Show the resultant image
	namedWindow("Contours", WINDOW_AUTOSIZE);
	imshow("Contours", drawing);
	waitKey(0);
}